cats/cats-sql/src/main/kotlin/com/netflix/spinnaker/cats/sql/cache/SqlCache.kt

```diff
@@ -649,123 +683,240 @@ class SqlCache(
       return result
     }
 
-    val toDelete = existingIds
+    val idHashesToDelete = existingIdHashes
       .asSequence()
-      .filter { !currentIds.contains(it) }
+      .filter { !currentIdHashes.contains(it) }
       .toSet()
 
-    evictAll(type, toDelete)
+    evictAllInternal(type, idHashesToDelete)
 
     return result
   }
 
-  private fun storeInformative(type: String, items: MutableCollection<CacheData>, cleanup: Boolean): StoreResult {
+  private fun storeInformative(
+    type: String,
+    items: MutableCollection<CacheData>,
+    cleanup: Boolean
+  ): StoreResult {
     val result = StoreResult()
 
-    val sourceAgents = items.filter { it.relationships.isNotEmpty() }
+    val relAgentHashToRelAgent = mutableMapOf<String, String>()
+    val relAgentHashes = items.asSequence()
+      .filter { it.relationships.isNotEmpty() }
       .map { it.relationships.keys }
       .flatten()
+      .map {
+        val relAgentHash = getAgentHash(it)
+        relAgentHashToRelAgent[relAgentHash] = it
```

Comment:
Maybe this is not weird in kotlin world, but feels a little strange to have this side effect as part of the map callback.